### PR TITLE
AcpiPlatformDxe: MADT force edge trigger mode

### PR DESCRIPTION
**Pull request by yamahata**
TDX doesn't support eoi intercept because cpu state is protected and
the way to inject interrupt into vcpu is only posted interrupt.
It implies level trigger interrupt can't be emulated and only edge
trigger is supported.
So forcibly report edge trigger for all legacy interrupt via interrupt
source override entry in MADT.

Signed-off-by: Isaku Yamahata <isaku.yamahata@intel.com>

### CODE REVIEW

**Comment by yamahata** [comment]
Alternative approach is for tdvf to copy interrupt source override from qemu instead of tdvf generating its own.

File: TdvfPkg/TdvfPkg.dec [hunk 1]
```diff
@@ -70,7 +70,7 @@
   ## Pcd8259LegacyModeEdgeLevel defines the default edge level for legacy
   #  mode's interrrupt controller.
   #  For the corresponding bits, 0 = Edge triggered and 1 = Level triggered.
-  gUefiTdvfPkgTokenSpaceGuid.Pcd8259LegacyModeEdgeLevel|0x0000|UINT16|0x20009
+  gUefiTdvfPkgTokenSpaceGuid.Pcd8259LegacyModeEdgeLevel|0x0000|UINT16|0x00000
```
Comment (mxu9) [inline TdvfPkg/TdvfPkg.dec]: 0x20009 is the token number of Pcd8259LegacyModeEdgeLevel. It should not be set to 0x00000

File: TdvfPkg/UefiPayload/AcpiPlatformDxe/AcpiPlatform.c [hunk 2]
```diff
@@ -74,17 +53,11 @@ InstallAcpiMadtTable (
...
-  // corresponds to the union of all possible interrupt assignments for the LNKA,
-  // LNKB, LNKC, LNKD PCI interrupt lines. See the DSDT.
-  //
-  PciLinkIsoCount = CountBits16 (PcdGet16 (Pcd8259LegacyModeEdgeLevel));
 
   NewBufferSize = 1                     * sizeof (*Madt) +
                   CpuCount              * sizeof (*LocalApic) +
                   1                     * sizeof (*IoApic) +
-                  (1 + PciLinkIsoCount) * sizeof (*Iso) +
+                  16 * sizeof (*Iso) +
```
Comment (mxu9) [inline TdvfPkg/UefiPayload/AcpiPlatformDxe/AcpiPlatform.c]: May I know why use the magic number 16 here? And please align the format

Comment (yamahata) [reply TdvfPkg/UefiPayload/AcpiPlatformDxe/AcpiPlatform.c]: The number of IRQ for intel ioapic 82093.
will fix the indent.

Comment (mxu9) [reply TdvfPkg/UefiPayload/AcpiPlatformDxe/AcpiPlatform.c]: Hi, Isaku
Do you have any update of the PR?

File: TdvfPkg/TdvfPkg.dec (hunk 1 above)
Comment (yamahata) [reply TdvfPkg/TdvfPkg.dec]: Why? The goal is to enforce all IRQ0-15 edge triggered.

File: TdvfPkg/UefiPayload/AcpiPlatformDxe/AcpiPlatform.c (hunk 2 above)
Comment (yamahata) [reply TdvfPkg/UefiPayload/AcpiPlatformDxe/AcpiPlatform.c]: I've updated it.

**Comment by yamahata** [comment]
BTW, TDVF(or OVMF) generates its own DSDT. and TDVF seems to have mostly same DSDT.
With TDX, interrupt must be edge triggered. So DSDT needs to be updated. interrupt routing and interrupt link device.
TDX-qemu has such changes. However TDVF simply ignores all of such DSDT changes.
What do  you think?

In practice, Linux uses MSI so that it doesn't matter, though.

File: TdvfPkg/TdvfPkg.dec (hunk 1 above)
Comment (mxu9) [reply TdvfPkg/TdvfPkg.dec]: I think there is a misunderstanding of the gUefiTdvfPkgTokenSpaceGuid.Pcd8259LegacyModeEdgeLevel|**0x0000**|UINT16|**0x20009** here.
0x20009 is the token number of Pcd8259LegacyModeEdgeLevel. 0x20009 itself should not be set to 0x00000.
The value of Pcd8259LegacyModeEdgeLevel is **0x0000**. It is correct.

Comment (yamahata) [reply TdvfPkg/TdvfPkg.dec]: Oh I see. thanks for explanation. I'll update the patch.

**Review by mxu9 — APPROVED** [review]

**Pull request merged**